src/Engine/UnitActionLifecycles/TerraformLifecycle.java

```diff
@@ -0,0 +1,128 @@
+import Units.Unit;
+
+public abstract class TerraformLifecycle
+{
+  public static class TerraformFactory extends UnitActionFactory
+  {
+    private static final long serialVersionUID = 1L;
+    public final TerrainType startType, endType;
+    public final String name;
+
+    public TerraformFactory(TerrainType typeFrom, TerrainType typeTo, String displayName)
+    {
+      startType = typeFrom;
+      endType = typeTo;
+      name = displayName;
+    }
+
+    @Override
+    public GameActionSet getPossibleActions(GameMap map, Path movePath, Unit actor, boolean ignoreResident)
+    {
+      XYCoord moveLocation = movePath.getEndCoord();
+      if( ignoreResident || map.isLocationEmpty(actor, moveLocation) )
+      {
+        if( startType == map.getLocation(moveLocation).getEnvironment().terrainType
+            && actor.materials > 0 )
+        {
+          return new GameActionSet(new TerraformAction(actor, movePath, this), false);
+        }
+      }
+      return null;
+    }
+
+    @Override
+    public String name()
+    {
+      return name;
+    }
+  } // ~Factory
+
+  public static class TerraformAction extends WaitLifecycle.WaitAction
+  {
+    private Unit actor = null;
+    private TerraformFactory type;
+
+    public TerraformAction(Unit unit, Path path, TerraformFactory pType)
+    {
+      super(unit, path);
+      actor = unit;
+      type = pType;
+    }
+
+    @Override
+    public GameEventQueue getEvents(MapMaster map)
+    {
+      // TERRAFORM actions consist of
+      //   MOVE
+      //   TERRAFORM
+      GameEventQueue terraformEvents = super.getEvents(map);
+
+      if( terraformEvents.size() < 1 ) // Fail out if invalid
+        return terraformEvents;
+      if( actor.materials < 1 ) // Fail out if we don't have materials
+        return terraformEvents;
+
+      GameEvent moveEvent = terraformEvents.peek();
+      if( !moveEvent.getEndPoint().equals(getMoveLocation()) ) // Fail out if pre-empted
+        return terraformEvents;
+
+      terraformEvents.add(new TerraformEvent(actor, getMoveLocation(), type.endType));
+
+      return terraformEvents;
+    }
+
+    @Override
+    public String toString()
+    {
+      return String.format("[Terraform %s to %s at %s with %s]", type.startType, type.endType, getMoveLocation(),
+          actor.toStringWithLocation());
+    }
+
+    @Override
+    public UnitActionFactory getType()
+    {
+      return type;
+    }
+  } // ~Action
+
+  public static class TerraformEvent extends MapChangeEvent
+  {
+    private Unit unit = null;
+    final int captureAmount;
+    final int priorCaptureAmount;
+
+    public TerraformEvent(Unit u, XYCoord loc, TerrainType endType)
+    {
+      super(loc, Environment.getTile(endType, Weathers.CLEAR));
+      unit = u;
+      XYCoord unitXY = new XYCoord(u.x, u.y);
+      priorCaptureAmount = (unitXY.equals(loc) ? unit.getCaptureProgress() : 0);
+      captureAmount = unit.getHP(); // TODO: Apply CO buffs.
```

Comment:
Do Rigs build more slowly when damaged on the carts too? I don't think I ever noticed that.

Reply:
They totally do. You probably noticed because they're high-priority targets :stuck_out_tongue:

Reply:
Nah, I noticed because I was reviewing code. I feel like I a) don't often build/see rigs, and b) rarely see them damaged without being quickly destroyed since they aren't front-line units.